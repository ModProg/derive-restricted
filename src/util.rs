//! Utility functions.

use core::iter::FromIterator;

use proc_macro2::Span;
use syn::{punctuated::Punctuated, Ident, Path, PathArguments, PathSegment, Token};

use crate::Item;

/// Create [`PathSegment`] from [`str`].
pub fn path_segment(ident: &str) -> PathSegment {
    PathSegment {
        ident: Ident::new(ident, Span::call_site()),
        arguments: PathArguments::None,
    }
}

/// Create [`Path`] from `[&str]`.
pub fn path(segments: &[&str]) -> Path {
    Path {
        leading_colon: Some(<Token![::]>::default()),
        segments: Punctuated::from_iter(segments.iter().map(|segment| path_segment(segment))),
    }
}

/// Return if given enum has any empty or unit variants. If not an enum, will always return `false`.
pub fn unit_found(data: &Item) -> bool {
    if let Item::Enum { variants, .. } = data {
        variants.iter().any(|variant| variant.fields.is_empty())
    } else {
        false
    }
}
