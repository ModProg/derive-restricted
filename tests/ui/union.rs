#[allow(unused_imports)]
use core::marker::PhantomData;

#[derive_restricted::derive_where(Clone)]
union Test<T> {
    field: PhantomData<T>,
}

fn main() {}
