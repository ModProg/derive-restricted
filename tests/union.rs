#![allow(clippy::clone_on_copy)]

mod util;

use derive_where::derive_where;

use self::util::{AssertClone, AssertCopy, Wrapper};

#[test]
fn single() {
	#[derive_where(Clone, Copy)]
	union Test<T> {
		a: Wrapper<T>,
	}

	let test_1 = Test { a: 42.into() };

	let _ = AssertClone(&test_1);
	let _ = AssertCopy(&test_1);

	let test_clone = test_1.clone();
	assert_eq!(unsafe { test_clone.a }, 42);

	let test_copy = test_1;
	assert_eq!(unsafe { test_copy.a }, 42);
}

#[test]
fn multiple() {
	#[derive_where(Clone, Copy)]
	union Test<T> {
		a: Wrapper<T>,
		b: Wrapper<T>,
		c: Wrapper<T>,
	}

	let test_1 = Test { a: 42.into() };
	let test_2 = Test { b: 43.into() };
	let test_3 = Test { c: 44.into() };

	let _ = AssertClone(&test_1);
	let _ = AssertCopy(&test_1);

	let test_clone = test_1.clone();
	assert_eq!(unsafe { test_clone.a }, 42);
	let test_clone = test_2.clone();
	assert_eq!(unsafe { test_clone.b }, 43);
	let test_clone = test_3.clone();
	assert_eq!(unsafe { test_clone.c }, 44);

	let test_copy = test_1;
	assert_eq!(unsafe { test_copy.a }, 42);
	let test_copy = test_2;
	assert_eq!(unsafe { test_copy.b }, 43);
	let test_copy = test_3;
	assert_eq!(unsafe { test_copy.c }, 44);
}
