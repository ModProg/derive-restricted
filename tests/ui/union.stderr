error: traits other then `Clone` and `Copy` aren't supported by unions
 --> tests/ui/union.rs:6:1
  |
6 | / union UnsupportedTrait<T> {
7 | |     a: Option<PhantomData<T>>,
8 | | }
  | |_^

error[E0277]: the trait bound `MissingCopy<T>: std::marker::Copy` is not satisfied
  --> tests/ui/union.rs:10:1
   |
10 | #[derive_where(Clone)]
   | ^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `MissingCopy<T>`
   |
note: required by a bound in `__AssertCopy`
  --> tests/ui/union.rs:10:1
   |
10 | #[derive_where(Clone)]
   | ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `__AssertCopy`
   = note: this error originates in the derive macro `::derive_where::DeriveWhere` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `MissingCopy<T>` with `#[derive(Copy)]`
   |
11 + #[derive(Copy)]
12 | union MissingCopy<T> {
   |
